README.md
# responsive_grid

Responsive Grid Layout and List for Flutter

### Responsive Grid Layout

With `ResponsiveGridRow` and `ResponsiveGridCol` you can get the same behavior of [bootstrap](https://getbootstrap.com) Grid System

Give every col the width it shall occupy at every size category assuming the total width is 12
    
```dart
        ResponsiveGridRow(
          children: [
            ResponsiveGridCol(
              lg: 12,
              child: Container(
                height: 100,
                alignment: Alignment(0, 0),
                color: Colors.purple,
                child: Text("lg : 12"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: Alignment(0, 0),
                color: Colors.green,
                child: Text("xs : 6 \r\nmd : 3"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: Alignment(0, 0),
                color: Colors.orange,
                child: Text("xs : 6 \r\nmd : 3"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: Alignment(0, 0),
                color: Colors.red,
                child: Text("xs : 6 \r\nmd : 3"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: Alignment(0, 0),
                color: Colors.blue,
                child: Text("xs : 6 \r\nmd : 3"),
              ),
            ),
          ],
        )

```

<img src="https://raw.githubusercontent.com/muhammad369/ResponsiveGrid_Flutter/master/images/1.jpg" width="300">   <img src="https://raw.githubusercontent.com/muhammad369/ResponsiveGrid_Flutter/master/images/2.jpg" height="300">


### Responsive Grid List

`ResponsiveGridList` works differently in that you only specify a desired width for the items and spacing, and it will decide how many items shall fit in a line, and the width of the item and spacing will change (only small change) to fill the entire width

```dart
    ResponsiveGridList(
        desiredItemWidth: 100,
        minSpacing: 10,
        children: [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20
        ].map((i) {
          return Container(
            height: 100,
            alignment: Alignment(0, 0),
            color: Colors.cyan,
            child: Text(i.toString()),
          );
        }).toList()
    )

```

<img src="https://raw.githubusercontent.com/muhammad369/ResponsiveGrid_Flutter/master/images/3.jpg" width="300">   <img src="https://raw.githubusercontent.com/muhammad369/ResponsiveGrid_Flutter/master/images/4.jpg" height="300">

### Responsive Staggered Grid List

The same as the `ResponsiveGridList`, but arranges items in columns inside one big row, so every item has independent height


<img src="https://raw.githubusercontent.com/muhammad369/ResponsiveGrid_Flutter/master/images/5.png" width="500">


### Utilities

`ResponsiveWidget` , `ResponsiveBuilder` and `responsiveValue()`

provide a custom responsive Widget, Builder or a single value, for every size tier

`ResponsiveLocalWidget` and `ResponsiveLocalBuilder`

provide a responsive Widget or Builder for its own width, not the viewport's, they are just wrappers on top of flutter `LayoutBuilder`

### Override breakpoints

Add before attach App widget 

```dart
ResponsiveGridBreakpoints.value = ResponsiveGridBreakpoints(
  xs: 600,
  sm: 905, 
  md: 1240,
  lg: 1440,
);
```


CHANGELOG.md
## [2.4.0] - Jun 18, 2023

* `ResponsiveLocalWidget` and `ResponsiveLocalBuilder`
provide a responsive Widget or Builder for its own width, not the viewport's, they are just wrappers on top of flutter `LayoutBuilder`

## [2.3.0] - Jun 8, 2023

* `ResponsiveStaggeredGridList`

## [2.2.0] - Mar 13, 2023

* `ResponsiveWidget` , `ResponsiveBuilder` and `responsiveValue()`
* some fixes from grateful PRs

## [2.1.0] - Apr 20, 2022

* Added `ResponsiveGridBreakpoints` for override default breakpoints
* Added optional `debug` property at `initScaling()` function
* Added `key` property at `ResponsiveGridRow`, `ResponsiveGridCol`, `ResponsiveGridList` widgets

## [2.0.0] - Apr 5, 2021

* Migrate to null safety

## [1.3.0] - Apr 4, 2021

* `ResponsiveGridRow` now has property `rowSegments`, 12 by default
instead of dividing the columns in a `ResponsiveGridRow` to 12 segments
you can now set the number of segments as you like

## [1.2.1] - Aug 23, 2020

* `ResponsiveGridList` uses ` ListView.builder` for better performance 

## [1.2.0] - Jun 20, 2020

* added property `ResponsiveGridRow.crossAxisAlignment` default value `CrossAxisAlignment.start`
* added property `ResponsiveGridList.rowMainAxisAlignment` default value `MainAxisAlignment.start`

## [1.1.1] - Nov 19, 2019

* (beta) use the scale() global function to scale font size, margin, padding, width, etc, but you must call initScaling() before

## [1.1.0] - Nov 9, 2019

* you can set the ResponsiveGridList to be rendered as a Column not a List (not scrollable), by setting 'scroll' to false

## [1.0.1] - Jul 28, 2019

* initial release.



example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:responsive_grid/responsive_grid.dart';

import '../lib/main.dart';

void main() {
  testWidgets('gridlist test', (WidgetTester tester) async {
    // test grid list
    await tester.pumpWidget(MyApp(homeWidget: _buildGridList()));

    for(int i = 1; i <= 20; i ++) {
      await tester.scrollUntilVisible(find.text('$i'), 10);
    }
  });

  testWidgets('gridlayout test', (WidgetTester tester) async {
    // test grid list
    await tester.pumpWidget(MyApp(homeWidget: _buildGridLayout()));

    // expect no exceptions

  });
}


Widget _buildGridList() {
  return ResponsiveGridList(
      rowMainAxisAlignment: MainAxisAlignment.start,

      desiredItemWidth: 100,
      //squareCells: true,
      minSpacing: 10,
      children: [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20
      ].map((i) {
        return Container(
          height: ((i%5) +1) * 100.0,
          alignment: Alignment(0, 0),
          color: Colors.cyan,
          child: Text(i.toString()),
        );
      }).toList());
}

Widget _buildGridLayout() {
  return Column(
    mainAxisAlignment: MainAxisAlignment.start,
    children: <Widget>[
      ResponsiveGridRow(
        children: [
          ResponsiveGridCol(
            lg: 12,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.purple,
              child: Text("lg : 12"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.green,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.orange,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.red,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.blue,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
        ],
      ),
    ],
  );
}

test/responsive_grid_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:responsive_grid/responsive_grid.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


example/flutter_app/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_app/main.dart';
import 'package:responsive_grid/responsive_grid.dart';

void main() {
  testWidgets('gridlist test', (WidgetTester tester) async {
    // test grid list
    await tester.pumpWidget(MyApp(homeWidget: _buildGridList()));

    for(int i = 1; i <= 20; i ++) {
      await tester.scrollUntilVisible(find.text('$i'), 10);
    }
  });

  testWidgets('gridlayout test', (WidgetTester tester) async {
    // test grid list
    await tester.pumpWidget(MyApp(homeWidget: _buildGridLayout()));

    // expect no exceptions

  });
}


Widget _buildGridList() {
  return ResponsiveGridList(
      rowMainAxisAlignment: MainAxisAlignment.start,

      desiredItemWidth: 100,
      //squareCells: true,
      minSpacing: 10,
      children: [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20
      ].map((i) {
        return Container(
          height: ((i%5) +1) * 100.0,
          alignment: Alignment(0, 0),
          color: Colors.cyan,
          child: Text(i.toString()),
        );
      }).toList());
}

Widget _buildGridLayout() {
  return Column(
    mainAxisAlignment: MainAxisAlignment.start,
    children: <Widget>[
      ResponsiveGridRow(
        children: [
          ResponsiveGridCol(
            lg: 12,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.purple,
              child: Text("lg : 12"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.green,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.orange,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.red,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
          ResponsiveGridCol(
            xs: 6,
            md: 3,
            child: Container(
              height: 100,
              alignment: Alignment(0, 0),
              color: Colors.blue,
              child: Text("xs : 6 \r\nmd : 3"),
            ),
          ),
        ],
      ),
    ],
  );
}

lib/responsive_grid.dart
library responsive_grid;

import 'package:flutter/widgets.dart';

//
// responsive grid layout
//

class ResponsiveGridBreakpoints {
  final double xs;
  final double sm;
  final double md;
  final double lg;
  final double xl;

  ResponsiveGridBreakpoints({
    this.xs = 576,
    this.sm = 768,
    this.md = 992,
    this.lg = 1200,
    this.xl = double.infinity,
  });

  static ResponsiveGridBreakpoints value = ResponsiveGridBreakpoints();
}

enum _GridTier { xs, sm, md, lg, xl }

_GridTier _currentSize(BuildContext context) {
  final breakpoints = ResponsiveGridBreakpoints.value;
  final mediaQueryData = MediaQuery.of(context);
  final width = mediaQueryData.size.width;

//  print(
//      "INFO orientation: ${mediaQueryData.orientation} , width: ${mediaQueryData.size.width}, height: ${mediaQueryData.size.height}");

  if (width < breakpoints.xs) {
    return _GridTier.xs;
  } else if (width < breakpoints.sm) {
    return _GridTier.sm;
  } else if (width < breakpoints.md) {
    return _GridTier.md;
  } else if (width < breakpoints.lg) {
    return _GridTier.lg;
  } else {
    // width >= 1200
    return _GridTier.xl;
  }
}

class ResponsiveGridRow extends StatelessWidget {
  final List<ResponsiveGridCol> children;
  final CrossAxisAlignment crossAxisAlignment;
  final int rowSegments;

  const ResponsiveGridRow({
    required this.children,
    this.crossAxisAlignment = CrossAxisAlignment.start,
    this.rowSegments = 12,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final rows = <Widget>[];

    int accumulatedWidth = 0;
    var cols = <Widget>[];

    for (var col in children) {
      var colWidth = col.currentConfig(context) ?? 1;
      //
      if (accumulatedWidth + colWidth > rowSegments) {
        if (accumulatedWidth < rowSegments) {
          cols.add(Spacer(
            flex: rowSegments - accumulatedWidth,
          ));
        }
        rows.add(Row(
          crossAxisAlignment: this.crossAxisAlignment,
          children: cols,
        ));
        // clear
        cols = <Widget>[];
        accumulatedWidth = 0;
      }
      //
      cols.add(col);
      accumulatedWidth += colWidth;
    }

    if (accumulatedWidth >= 0) {
      if (accumulatedWidth < rowSegments) {
        cols.add(Spacer(
          flex: rowSegments - accumulatedWidth,
        ));
      }
      rows.add(Row(
        crossAxisAlignment: crossAxisAlignment,
        children: cols,
      ));
    }

    return Column(
      children: rows,
    );
  }
}

class ResponsiveGridCol extends StatelessWidget {
  final _config = <int?>[]..length = 5;
  final Widget child;

  ResponsiveGridCol({
    int xs = 12,
    int? sm,
    int? md,
    int? lg,
    int? xl,
    required this.child,
    Key? key,
  }) : super(key: key) {
    _config[_GridTier.xs.index] = xs;
    _config[_GridTier.sm.index] = sm ?? _config[_GridTier.xs.index];
    _config[_GridTier.md.index] = md ?? _config[_GridTier.sm.index];
    _config[_GridTier.lg.index] = lg ?? _config[_GridTier.md.index];
    _config[_GridTier.xl.index] = xl ?? _config[_GridTier.lg.index];
  }

  int? currentConfig(BuildContext context) {
    return _config[_currentSize(context).index];
  }

  @override
  Widget build(BuildContext context) {
    return Expanded(
      flex: currentConfig(context) ?? 1,
      child: child,
    );
  }
}

//
// responsive grid list
//

class ResponsiveGridList extends StatelessWidget {
  final double desiredItemWidth, minSpacing;
  final List<Widget> children;
  final bool squareCells, scroll;
  final MainAxisAlignment rowMainAxisAlignment;
  final bool shrinkWrap;
  final ScrollController? controller;
  final ScrollPhysics? physics;

  const ResponsiveGridList({
    required this.desiredItemWidth,
    this.minSpacing = 1,
    this.squareCells = false,
    this.scroll = true,
    required this.children,
    this.rowMainAxisAlignment = MainAxisAlignment.start,
    this.shrinkWrap = false,
    this.controller,
    this.physics,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        if (children.isEmpty) return Container();

        double width = constraints.maxWidth;

        double N = (width - minSpacing) / (desiredItemWidth + minSpacing);

        int n;
        double spacing, itemWidth;

        if (N % 1 == 0) {
          n = N.floor();
          spacing = minSpacing;
          itemWidth = desiredItemWidth;
        } else {
          if (N >= 2) {
            n = N.floor();
          } else {
            n = 1;
          }

          double dw = width - (n * (desiredItemWidth + minSpacing) + minSpacing);

          itemWidth = desiredItemWidth + (dw / n) * (desiredItemWidth / (desiredItemWidth + minSpacing));

          spacing = (width - itemWidth * n) / (n + 1);
        }

        if (scroll) {
          return ListView.builder(
              controller: controller,
              physics: physics,
              shrinkWrap: shrinkWrap,
              itemCount: (children.length / n).ceil() * 2 - 1,
              itemBuilder: (context, index) {
                //if (index * n >= children.length) return null;
                //separator
                if (index % 2 == 1) {
                  return SizedBox(
                    height: minSpacing,
                  );
                }
                //item
                final rowChildren = <Widget>[];
                index = index ~/ 2;
                for (int i = index * n; i < (index + 1) * n; i++) {
                  if (i >= children.length) break;
                  rowChildren.add(children[i]);
                }
                return _ResponsiveGridListRow(
                  mainAxisAlignment: rowMainAxisAlignment,
                  itemWidth: itemWidth,
                  spacing: spacing,
                  squareCells: squareCells,
                  children: rowChildren,
                );
              });
        } else {
          final rows = <Widget>[];
          rows.add(SizedBox(
            height: minSpacing,
          ));
          //
          for (int j = 0; j < (children.length / n).ceil(); j++) {
            final rowChildren = <Widget>[];
            //
            for (int i = j * n; i < (j + 1) * n; i++) {
              if (i >= children.length) break;
              rowChildren.add(children[i]);
            }
            //
            rows.add(_ResponsiveGridListRow(
              mainAxisAlignment: rowMainAxisAlignment,
              itemWidth: itemWidth,
              spacing: spacing,
              squareCells: squareCells,
              children: rowChildren,
            ));

            rows.add(SizedBox(
              height: minSpacing,
            ));
          }

          return Column(
            children: rows,
          );
        }
      },
    );
  }
}

class _ResponsiveGridListRow extends StatelessWidget {
  final double spacing, itemWidth;
  final List<Widget> children;
  final bool squareCells;
  final MainAxisAlignment mainAxisAlignment;

  const _ResponsiveGridListRow({
    required this.itemWidth,
    required this.spacing,
    required this.squareCells,
    required this.children,
    this.mainAxisAlignment = MainAxisAlignment.start,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisAlignment: mainAxisAlignment,
      children: _buildChildren(),
    );
  }

  List<Widget> _buildChildren() {
    final list = <Widget>[];

    list.add(SizedBox(
      width: spacing,
    ));

    for (var child in children) {
      list.add(SizedBox(
        width: itemWidth,
        height: squareCells ? itemWidth : null,
        child: child,
      ));
      list.add(SizedBox(
        width: spacing,
      ));
    }

    return list;
  }
}

//
// ResponsiveStaggeredGridList
//

class ResponsiveStaggeredGridList extends StatelessWidget {
  final double desiredItemWidth, minSpacing;
  final List<Widget> children;
  final CrossAxisAlignment crossAxisAlignment;
  final ScrollController? controller;
  final ScrollPhysics? physics;
  final bool scroll;

  const ResponsiveStaggeredGridList({
    required this.desiredItemWidth,
    this.minSpacing = 1,
    required this.children,
    this.crossAxisAlignment = CrossAxisAlignment.start,
    this.controller,
    this.physics,
    this.scroll = true,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        if (children.isEmpty) return Container();

        double width = constraints.maxWidth;

        double N = (width - minSpacing) / (desiredItemWidth + minSpacing);

        int n;
        double spacing, itemWidth;

        if (N % 1 == 0) {
          n = N.floor();
          spacing = minSpacing;
          itemWidth = desiredItemWidth;
        } else {
          if (N >= 2) {
            n = N.floor();
          } else {
            n = 1;
          }

          double dw = width - (n * (desiredItemWidth + minSpacing) + minSpacing);

          itemWidth = desiredItemWidth + (dw / n) * (desiredItemWidth / (desiredItemWidth + minSpacing));

          spacing = (width - itemWidth * n) / (n + 1);
        }

        final List<List<Widget>> colsChildren = [];
        colsChildren.addAll(List.generate(
            n,
            (index) => <Widget>[
                  SizedBox(
                    height: spacing,
                  )
                ]));

        //
        for (int j = 0; j < children.length; j++) {
          final index = j % n;
          colsChildren[index].add(children[j]);
          colsChildren[index].add(SizedBox(
            height: spacing,
          ));
        }

        final cols = <Widget>[];

        cols.add(SizedBox(
          width: spacing,
        ));
        //
        for (int i = 0; i < colsChildren.length; i++) {
          cols.add(SizedBox(
            width: itemWidth,
            child: Column(
              children: colsChildren[i],
            ),
          ));
          cols.add(SizedBox(
            width: spacing,
          ));
        }

        return scroll
            ? SingleChildScrollView(
                physics: physics,
                child: Row(
                  crossAxisAlignment: crossAxisAlignment,
                  children: cols,
                ),
              )
            : Row(
                crossAxisAlignment: crossAxisAlignment,
                children: cols,
              );
      },
    );
  }
}

//
// Utilities
//

/// a widget for certain tier applies also for larger tiers unless overridden, so you must set xs at least
class ResponsiveWidget extends StatelessWidget {
  final Widget? sm, md, lg, xl;
  final Widget xs;

  const ResponsiveWidget({Key? key, this.lg, this.md, this.sm, this.xl, required this.xs}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    var w = MediaQuery.of(context).size.width;
    if (w >= ResponsiveGridBreakpoints.value.lg && xl != null) {
      return xl!;
    }
    if (w >= ResponsiveGridBreakpoints.value.md && lg != null) {
      return lg!;
    }
    if (w >= ResponsiveGridBreakpoints.value.sm && md != null) {
      return md!;
    }
    if (w >= ResponsiveGridBreakpoints.value.xs && sm != null) {
      return sm!;
    }
    return xs;
  }
}

/// a builder for certain tier applies also for larger tiers, so you must set xs at least
class ResponsiveBuilder extends StatelessWidget {
  final Widget child;
  final Function(BuildContext context, Widget child)? sm, md, lg, xl;
  final Function(BuildContext context, Widget child) xs;

  const ResponsiveBuilder({
    Key? key,
    required this.child,
    required this.xs,
    this.sm,
    this.md,
    this.lg,
    this.xl,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    var w = MediaQuery.of(context).size.width;
    if (w >= ResponsiveGridBreakpoints.value.lg && xl != null) {
      return xl!(context, child);
    }
    if (w >= ResponsiveGridBreakpoints.value.md && lg != null) {
      return lg!(context, child);
    }
    if (w >= ResponsiveGridBreakpoints.value.sm && md != null) {
      return md!(context, child);
    }
    if (w >= ResponsiveGridBreakpoints.value.xs && sm != null) {
      return sm!(context, child);
    }
    return xs(context, child);
  }
}

/// a value for certain tier applies also for larger tiers unless overridden, so you must set xs at least
T responsiveValue<T>(BuildContext context, {required T xs, T? sm, T? md, T? lg, T? xl}) {
  var w = MediaQuery.of(context).size.width;
  if (w >= ResponsiveGridBreakpoints.value.lg && xl != null) {
    return xl;
  }
  if (w >= ResponsiveGridBreakpoints.value.md && lg != null) {
    return lg;
  }
  if (w >= ResponsiveGridBreakpoints.value.sm && md != null) {
    return md;
  }
  if (w >= ResponsiveGridBreakpoints.value.xs && sm != null) {
    return sm;
  }
  return xs;
}

// local responsive

class ResponsiveWidgetConfig{
  final double upToWidth;
  final Widget child;

  ResponsiveWidgetConfig({required this.upToWidth, required this.child});

}

class ResponsiveBuilderConfig{
  final double upToWidth;
  final Function(BuildContext context, Widget child) builder;

  ResponsiveBuilderConfig({required this.upToWidth, required this.builder});

}

class ResponsiveLocalWidget extends StatelessWidget{

  final List<ResponsiveWidgetConfig> configs;

  /// a widget that changes according to its own width,
  /// the configs are assumed to be provided in ascending order
  ResponsiveLocalWidget({Key? key, required this.configs}) : super(key: key){
    assert(configs.isNotEmpty);
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints){
      for(var config in configs){
        if(constraints.maxWidth <= config.upToWidth) {
          return config.child;
        }
      }
      //
      return configs.last.child;
    });
  }

}


class ResponsiveLocalBuilder extends StatelessWidget{

  final List<ResponsiveBuilderConfig> configs;
  final Widget child;

  /// a widget that changes according to its own width,
  /// the configs are assumed to be provided in ascending order
  ResponsiveLocalBuilder({Key? key, required this.configs, required this.child}) : super(key: key){
    assert(configs.isNotEmpty);
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints){
      for(var config in configs){
        if(constraints.maxWidth <= config.upToWidth) {
          return config.builder(context, child);
        }
      }
      //
      return configs.last.builder(context, child);
    });
  }

}


class ResponsiveLayoutBuilderConfig{
  final double upToWidth;
  final Function(BuildContext context, List<Widget> children) builder;

  ResponsiveLayoutBuilderConfig({required this.upToWidth, required this.builder});

}

class ResponsiveLocalLayoutBuilder extends StatelessWidget{

  final List<ResponsiveLayoutBuilderConfig> configs;
  final List<Widget> children;

  /// a widget that changes according to its own width,
  /// the configs are assumed to be provided in ascending order
  ResponsiveLocalLayoutBuilder({Key? key, required this.configs, required this.children}) : super(key: key){
    assert(configs.isNotEmpty);
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints){
      for(var config in configs){
        if(constraints.maxWidth <= config.upToWidth) {
          return config.builder(context, children);
        }
      }
      //
      return configs.last.builder(context, children);
    });
  }

}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:responsive_grid/responsive_grid.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  final Widget? homeWidget;

  const MyApp({
    this.homeWidget,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: homeWidget ?? const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: _buildResponsiveLocalBuilder());
  }

  Widget _buildGridList() {
    return ResponsiveGridList(
        rowMainAxisAlignment: MainAxisAlignment.center,
        desiredItemWidth: 100,
        minSpacing: 10,
        children: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20].map((i) {
          return Container(
            height: ((i % 5) + 1) * 100.0,
            alignment: const Alignment(0, 0),
            color: Colors.cyan,
            child: Text(i.toString()),
          );
        }).toList());
  }

  Widget _buildGridLayout() {
    return Column(
      mainAxisAlignment: MainAxisAlignment.start,
      children: <Widget>[
        ResponsiveGridRow(
          children: [
            ResponsiveGridCol(
              lg: 12,
              child: Container(
                height: 100,
                alignment: const Alignment(0, 0),
                color: Colors.purple,
                child: const Text("lg : 12"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: const Alignment(0, 0),
                color: Colors.green,
                child: const Text("xs : 6 \r\nmd : 3"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: const Alignment(0, 0),
                color: Colors.orange,
                child: const Text("xs : 6 \r\nmd : 3"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: const Alignment(0, 0),
                color: Colors.red,
                child: const Text("xs : 6 \r\nmd : 3"),
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              md: 3,
              child: Container(
                height: 100,
                alignment: const Alignment(0, 0),
                color: Colors.blue,
                child: const Text("xs : 6 \r\nmd : 3"),
              ),
            ),
          ],
        ),
      ],
    );
  }

  Widget _buildGridLayout_testCrossAlign() {
    return Column(
      mainAxisAlignment: MainAxisAlignment.start,
      children: <Widget>[
        ResponsiveGridRow(
          crossAxisAlignment: CrossAxisAlignment.end,
          children: [
            ResponsiveGridCol(
              xs: 6,
              child: Column(
                children: [
                  Container(
                    height: 100,
                    color: Colors.blue,
                    margin: const EdgeInsets.all(10),
                  ), // height 100px
                  Container(
                    height: 100,
                    color: Colors.blueGrey,
                    margin: const EdgeInsets.all(10),
                  ) // height 100px
                ],
              ),
            ),
            ResponsiveGridCol(
              xs: 6,
              child: Container(
                height: 400,
                color: Colors.black45,
                margin: const EdgeInsets.all(10),
              ), // height 500px
            )
          ],
        ),
      ],
    );
  }

  Widget _resposiveWidgetTest() {
    return Center(
        child: const ResponsiveWidget(xs: Text('xs'), sm: Text('sm'), md: Text('md'), lg: Text('lg'), xl: Text('xl')));
  }

  Widget _resposiveValueTest() {
    return Center(child: Text(responsiveValue(context, xs: 'xs', lg: 'lg', md: 'md', sm: 'sm', xl: 'xl')));
  }

  Widget _resposiveBuilderTest() {
    return Center(
        child: ResponsiveBuilder(
            child: const Text(
              'test text',
              style: TextStyle(fontSize: 30),
            ),
            xs: (_, child) => Container(
                  alignment: Alignment.topRight,
                  child: child,
                ),
            sm: (_, child) => Container(
                  alignment: Alignment.topLeft,
                  child: child,
                ),
            md: (_, child) => Container(
                  alignment: Alignment.center,
                  child: child,
                ),
            lg: (_, child) => Container(
                  alignment: Alignment.bottomRight,
                  child: child,
                ),
            xl: (_, child) => Container(
                  alignment: Alignment.bottomLeft,
                  child: child,
                )));
  }

  Widget _buildStaggeredGridList() {
    return ResponsiveStaggeredGridList(
        //crossAxisAlignment: CrossAxisAlignment.end,
        desiredItemWidth: 100,
        minSpacing: 10,
        children: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20].map((i) {
          return Container(
            height: ((i % 3) + 1) * 100.0,
            alignment: const Alignment(0, 0),
            color: Colors.cyan,
            child: Text(i.toString()),
          );
        }).toList());
  }

  Widget _buildResponsiveLocalBuilder() {
    return ResponsiveLocalLayoutBuilder(
      children: [Container(
        width: 100,
        height: 100,
        color: Colors.blue,
      )],
      configs: [500, 600, 700, 900]
          .map((e) => ResponsiveLayoutBuilderConfig(
              upToWidth: e.toDouble(),
              builder: ( context,  children) =>
                      Column(
                    children: children)
                  ))
          .toList(),
    );
  }
}
